src/DataStructures/Index.hpp

```diff
@@ -118,12 +141,75 @@ class Index {
 /// Index. Note that the first dimension of the Index varies fastest when
 /// computing the collapsed index.
 template <size_t N>
-size_t collapsed_index(const Index<N>& index, const Index<N>& extents) noexcept;
+size_t collapsed_index(const Index<N>& index,
+                       const Index<N>& extents) noexcept {
+  return extents.collapsed_index(index.indices_);
+}
 
 template <size_t N>
 std::ostream& operator<<(std::ostream& os, const Index<N>& i);
 
 /// \cond HIDDEN_SYMBOLS
+#ifdef SPECTRE_DEBUG
+template <size_t Dim>
+void Index<Dim>::collapsed_index_check(
+    const std::array<size_t, Dim>& index) const noexcept {
+  for (size_t d = 0; d < Dim; ++d) {
+    ASSERT(gsl::at(index, d) < gsl::at(indices_, d),
+           "The requested index in the dimension "
+               << d << " with value " << gsl::at(index, d)
+               << " exceeds the number of grid "
+                  "points "
+               << gsl::at(indices_, d));
+  }
+}
+#endif
+
+template <>
+SPECTRE_ALWAYS_INLINE size_t
```

Comment:
Please add a comment why you prefer this implementation over the previous, so that it won't be accidentally changed back in the future. My guess is that you want it to `inline` better?

Reply:
Also, why do these need to be in the header? Is that also for inlining?